homeassistant/components/notify/html5.py

```diff
@@ -129,7 +133,84 @@ def post(self, request):
             return self.json_message('Error saving registration.',
                                      HTTP_INTERNAL_SERVER_ERROR)
 
-        return self.json_message("Push notification subscriber registered.")
+        return self.json_message('Push notification subscriber registered.')
+
+
+class HTML5PushCallbackView(HomeAssistantView):
+    """Accepts push registrations from a browser."""
+
+    requires_auth = False
+    url = '/api/notify.html5/callback'
+    name = 'api:notify.html5/callback'
+
+    def __init__(self, hass, json_path):
+        """Init HTML5PushCallbackView."""
+        super().__init__(hass)
+        self.json_path = json_path
+
+    def decode_jwt(self, token):
+        """Find the registration that signed this JWT and return it."""
+        import jwt
+        registrations = _load_config(self.json_path)
+        for reg in registrations.values():
+            auth_key = reg['subscription']['keys']['auth']
+            try:
+                payload = jwt.decode(token, auth_key)
+            except jwt.exceptions.DecodeError:
+                continue
+            # pylint: disable=broad-except
+            except Exception as jwt_decode_error:
+                return jwt_decode_error
+            else:
+                return payload
+
+    # The following is based on code from Auth0
+    # https://auth0.com/docs/quickstart/backend/python
+    def json_error(self, code, description):
+        """Return a JSON formatted forbidden error."""
+        return self.json({'code': code, 'description': description},
+                         status_code=HTTP_UNAUTHORIZED)
+
+    # pylint: disable=too-many-return-statements
+    def check_authorization_header(self, request):
+        """Check the authorization header."""
+        import jwt
+        auth = request.headers.get('Authorization', None)
+        if not auth:
+            return self.json_error('authorization_header_missing',
+                                   'Authorization header is expected')
+
+        parts = auth.split()
+
+        if parts[0].lower() != 'bearer':
+            return self.json_error('invalid_header',
+                                   'Authorization header must '
+                                   'start with Bearer')
+        elif len(parts) == 1:
+            return self.json_error('invalid_header',
+                                   'Token not found')
+        elif len(parts) > 2:
+            return self.json_error('invalid_header',
+                                   'Authorization header must be Bearer token')
+
+        token = parts[1]
+        try:
+            self.decode_jwt(token)
```

Comment:
If `decode_jwt` runs through all authorizations and only runs into `DecodeError`, it will never hit a return statement.